dali/pipeline/graph/op_graph2.h

```diff
@@ -0,0 +1,222 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_GRAPH_OP_GRAPH2_H_
+#define DALI_PIPELINE_GRAPH_OP_GRAPH2_H_
+
+#include <cassert>
+#include <list>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <vector>
+#include <utility>
+#include "dali/core/common.h"
+#include "dali/pipeline/operator/op_spec.h"
+
+namespace dali {
+namespace graph {
+
+class OpGraph;
+
+struct OpNode;
+struct DataNode;
+
+using OpNodeList = std::list<OpNode>;
+using DataNodeList = std::list<DataNode>;
+
+struct OpNode {
+  /** The iterator that points to this object in the enclosing list. */
+  OpNodeList::iterator iter{};
+
+  OpNode(std::string instance_name, OpSpec spec)
+  : instance_name(std::move(instance_name)), spec(std::move(spec)) {}
+
+  /** A visit marker for various graph processing algorithms. */
+  mutable bool visited = false;
+  /** A visit marker for cycle detection. */
+  mutable bool visit_pending = false;
```

Comment:
I know.... on the other hand, there's no way to implement it that's efficient and the elegance of the alternatives is also questionable. I know we're not writing a compiler and the size of the graph is going to be small to moderate (thousands of nodes, not millions), but even then looking up helper structures for _extremely common_ type of metadata, such as visit markers, seems silly.
Another option:
```C++
union {
    void *meta_ptr;
    uint64_t meta_int;
    uint8_t meta_bytes[8];
};
```
If an algorithm needs more then 8 bytes, it can create a list of metadata items and use the meta_ptr to associate it with the nodes.